### ISSUE TITLE
FR : Make face used for line highlighting in biblio-selection-mode customizable

### ISSUE THREAD

**Issue by aagon** [cpitclaudel/biblio.el#43]
For the moment, defining a specific face and trying to set it as buffer local for `hl-line-face` in the `biblio-selection-mode-hook` does not work : the face used by `hl-line-mode` is still `hl-line` when we arrive in the `biblio-selection-mode` buffer.

It only works if the buffer-local setting is made in `buffer-selection-mode` **before** `hl-line-mode` is activated.

I propose the following : 
- Create a face specific for line highlighting, like this : 
```elisp
(defface biblio-highlight-extend-face `((t (:inherit highlight
						     ,@(and (>= emacs-major-version 27) '(:extend t)))))
  "Face used for highlighting lines."
  :group 'biblio-faces)
```
It will inherit the highlight face, set by most themes (hence always with a high contrast to the non highlighted background), but will still be easily customizable.
- Set this face as the buffer-local face for `hl-line-face` in the definition of the minor mode, as such : 
```elisp
(define-derived-mode biblio-selection-mode fundamental-mode biblio--selection-mode-name-base
  "Browse bibliographic search results.
\\{biblio-selection-mode-map}"
  (set (make-local-variable 'hl-line-face) 'biblio-highlight-extend-face)
  (hl-line-mode 1)
  (visual-line-mode)
  (setq-local truncate-lines nil)
  (setq-local cursor-type nil)
  (setq-local buffer-read-only t)
  (setq-local mode-name '(:eval (biblio--selection-mode-name)))
  (setq-local
   header-line-format
   `(:eval
     (concat
      (ignore-errors
        (propertize " " 'display '(space :align-to 0) 'face 'fringe))
      (substitute-command-keys
       (biblio-join "   "
         "\\[biblio--selection-help]: Help"
         "\\[biblio--selection-insert],\\[biblio--selection-insert-quit]: Insert BibTex"
         "\\[biblio--selection-copy],\\[biblio--selection-copy-quit]: Copy BibTeX"
         "\\[biblio--selection-extended-action]: Extended action"
         "\\[biblio--selection-browse]: Open in browser"
         "\\[biblio--selection-change-buffer]: Change buffer"))))))
```
This produces clearly visible and theme-consistent highlighting, regardless of theme.

If you agree, I can prepare the trivial PR with the changes.

Thanks again for this wonderful package,

Aymeric Agon-Rambosson

**Comment by cpitclaudel** [comment]
:+1 Sounds good to me.  Shouldn't we inherit from hl-line instead of highlight?  (and won't we inherit `:extend` from it?  Of if it doesn't have `:extend` then why do we need it?)

Also, I prefer setq-local over `set` + `make-local-variable`

**Comment by aagon** [comment]
>  Shouldn't we inherit from hl-line instead of highlight?

In stock emacs, hl-line inherits from highlight. But some themes (I am thinking of doom-themes, for instance) set `hl-line` to a background with low contrast to the normal one, because it is meant to be used in a context where the cursor is still visible, and `highlight` to a background with high contrast to the normal one, because it is meant to be used in a context where the cursor is hidden, like `biblio-selection-mode`.

The reason I actually looked into this is because with most doom-themes (I am using molokai), the contrast provided by `hl-line` is nigh invisible, making impossible to see which entry is selected in `biblio-selection-mode`, whereas the one provided by `highlight` is perfect. And I noticed that changing the face required modifying the definition of the minor mode itself (I wouldn't have bothered you otherwise).

Because we inherit directly from `highlight` and not `hl-line`, I prefer to inherit explicitely `:extend` (like `hl-line` does, BTW).

> Also, I prefer `setq-local` over `set` + `make-local-variable`

Done

> (It would be great to add a test for this, but it's not super high priority)

What would a test look like in this context ? I could make a trivial test to check whether the face is actually the one we want in `biblio-selection-mode`, but I'm not sure how I would test it myself (I don't know what "cask" or "buttercup" are, I run linux).

```elisp
(it "has the right major mode"
            (with-current-buffer results-buffer
              (expect major-mode :to-equal 'biblio-selection-mode)))
;; New test :
(it "sets the right face"
            (with-current-buffer results-buffer
              (expect (get-char-property (point) 'face) :to-equal 'biblio-highlight-extend-face)))
```
Does that seem OK ?

**Comment by aagon** [comment]
>  Shouldn't we inherit from hl-line instead of highlight?

Stock emacs makes `hl-line` inherit from `highlight`, but some themes (doom-themes in particular) set the background of each face differently : `hl-line` with low contrast to the normal background, to be used in an environment where the cursor is still visible, and `highlight` with high contrast to the normal background, perfect for environments where the cursor is hidden and the line highlighting is the only clue as to where the point is, as in `biblio-selection-mode`. In fact, I started this issue because with my theme (doom-molokai), the selected biblio entry was nigh-indistinguishable from the others, and setting a new face required modifying the minor mode.

So IMO best to inherit from `highlight` than from `hl-line` : doesn't change a thing for stock, but behaves better with some themes. And because we inherit from `highlight`, explicit `:extend` makes sense (I wanted to make the inheritance from `highlight` as similar as the of `hl-line` from `highlight`, which explicitly uses `:extend`).

> Also, I prefer `setq-local` over `set` + `make-local-variable`

Done.

> (It would be great to add a test for this, but it's not super high priority)

The only test I can think of would be to check whether the face is the right one in `biblio-selection-mode`, like this : 
```elisp
(it "has the right major mode"
            (with-current-buffer results-buffer
              (expect major-mode :to-equal 'biblio-selection-mode)))
;; New test :
(it "has the right face"
            (with-current-buffer results-buffer
              (expect (face-at-point) :to-equal 'biblio-highlight-extend-face)))
```
Let me know if you agree with this.

**Comment by cpitclaudel** [comment]
All of this makes complete sense.  Thanks a lot.

**Comment by aagon** [comment]
Closed by #44
